doc/adr/0002-steppable-cek-machine.md

```diff
@@ -0,0 +1,235 @@
+# ADR 2: Steppable CEK machine
+
+Date: 2022-10
+
+## Authors
+
+Marty Stumpf <marty.stumpf@iohk.io>  
+Ziyang Liu <ziyang.liu@iohk.io>
+
+## Status
+
+Proposed
+
+## Context
+
+In order to have a minimal viable product of a debugger for Plutus, we need a CEK machine that will give us more information for debugging than our current one.
+
+Implementing the steppable CEK machine is a non-trivial task and involves some design decisions:
+
+- should the debugging machine be a separate one?
+- should we modify the existing CEK machine to support stepping, or implement something entirely new?
+- how do we ensure performance and conformance?
+
+## Decision: A separate machine
+
+We came to the conclusion that the debugging machine should be a separate one. The main reason is that we don't want to slow down the production machine.
+
+## Argument: A separate machine
+
+The debugging machine will give the most accurate information if it was the one in production. However, the drawback of that is that we have to care about its performance.
```

Comment:
The idea is that if the production machine was the debugging machine also, it certainly knew what it just did, and could give us that same information. But with a separate machine, this is not as guaranteed. I will make it more clear. This is along the same line of avoiding making mistakes.

Reply:
I read "it will give the most accurate information" as "even if we implement it correctly, it might somehow give less accurate (?) information", rather than "if we do it this way there is more scope for us to implement it wrongly and have bugs".

I also think it's a bit misleading to say that in the alternative where we share code they are the _same machine_. IMO they are two machines that _share code_. So I would say something like "if we share code between the steppable machine an the production machine there is less scope for us to make mistakes".

Reply:
If we were C programmers we might keep the two machines synchronised by using a shell script to extract relevant code from the real evaluator and then pasting it into the source for the debugging machine using `#include`.  Of course, we're far too principled to do anything like that ................. aren't we?

Reply:
Oh, we can do that in Haskell too with enough abuse of CPP :D

Reply:
> Oh, we can do that in Haskell too with enough abuse of CPP :D

Yes, I was trying to suggest that without saying it out loud!